api/service/service.go

```diff
@@ -0,0 +1,106 @@
+package service
+
+import (
+	"context"
+	uuid "github.com/satori/go.uuid"
+	"fmt"
+	"io"
+	"time"
+	"encoding/json"
+
+	"github.com/spaceuptech/space-api-go/api/proto"
+	"github.com/spaceuptech/space-api-go/api/config"
+	"github.com/spaceuptech/space-api-go/api/model"
+	"github.com/spaceuptech/space-api-go/api/utils"
+)
+
+// Service contains the methods for the service instance
+type Service struct {
+	config  *config.Config
+	service string
+	id      string
+	funcs   map[string]func(*model.Message, *model.Message, func(string, interface{}))
+}
+
+func Init(config *config.Config, service string) *Service {
+	id := uuid.NewV1().String()
+	return &Service{config, service, id, make(map[string]func(*model.Message, *model.Message, func(string, interface{})))}
+}
+
+func (s *Service) RegisterFunc(funcName string, function func(*model.Message, *model.Message, func(string, interface{}))) {
+	s.funcs[funcName] = function
+}
+
+func (s *Service) Start() {
+	registerRequest := &proto.FunctionsPayload{Service: s.service, Type: utils.TypeServiceRegister, Id: s.id, Project: s.config.Project, Token: s.config.Token}
+	var toClose bool = false
+	for {
+		waitc := make(chan struct{})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		stream, err := s.config.Transport.GetStub().Service(ctx)
+		if err != nil {
+			fmt.Println("Unable to connect...")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		go func() {
+			for {
+				in, err := stream.Recv()
+				if err == io.EOF {
+					close(waitc)
+					toClose = true
+					return
+				}
+				if err != nil {
+					fmt.Println("Failed to receive a request:", err)
+					close(waitc)
+					return
+				} else {
+					if in.Type == utils.TypeServiceRegister {
+						if in.Id == s.id {
+							temp := make(map[string]bool)
+							json.Unmarshal(in.Params, &temp)
+							if temp["ack"] {
+								fmt.Println("Service registered")
+							} else {
+								panic("Could Not Register Service")
+							}
+						}
+					} else if in.Type == utils.TypeServiceRequest {
+						function, found := s.funcs[in.Function]
+						if found {
+							function(&model.Message{in.Params}, &model.Message{in.Auth}, func(kind string, res interface{}) {
```

Comment:
You should call this function in a goroutine

Reply:
Ok